mne/coreg.py

```diff
@@ -1223,3 +1227,586 @@ def _scale_xfm(subject_to, xfm_fname, mri_name, subject_from, scale,
                 F_mri_ras, 'ras', 'ras'),
             F_ras_mni, 'ras', 'mni_tal')
     _write_fs_xfm(fname_to, T_ras_mni['trans'], kind)
+
+
+class Coregistration(object):
+    """
+    Class for MRI<->MEG coregistration.
+
+    Parameters
+    ----------
+    info : instance of Info
+        The measurement info.
+    subject : str
+        Name of the subject the data are defined for.
+    subjects_dir : path-like
+        Path to MRI subjects directory.
+
+    Notes
+    -----
+    Internal computation quantities ``parameters`` are in units of (for X/Y/Z):
+    - ``parameters[:3]`` are in radians
+    - ``parameters[3:6]`` are in m
+    - ``paramteres[6:9]`` are in scale proportion
+    """
+
+    def __init__(self, info, subject, subjects_dir):
+        self._default_parameters = [0., 0., 0., 0., 0., 0., 1., 1., 1.]
+        self._parameters = list(self._default_parameters)
+        self._last_parameters = list(self._default_parameters)
+
+        self._n_scale_params = (0, 1, 3)
+        self._n_scale_param = self._n_scale_params[0]
+        self._coord_frames = ('mri', 'head')
+        self._coord_frame = self._coord_frames[0]
+        self._icp_iterations = 20
+        self._icp_angle = 0.2
+        self._icp_distance = 0.2
+        self._icp_scale = 0.2
+        self._icp_fid_matches = ('nearest', 'matched')
+        self._icp_fid_match = self._icp_fid_matches[0]
+        self._grow_hair = 0.
+        self._lpa_weight = 1.
+        self._nasion_weight = 10.
+        self._rpa_weight = 1.
+        self._hsp_weight = 1.
+        self._eeg_weight = 1.
+        self._hpi_weight = 1.
+
+        self._hsp = _DigSource(info)
+        self._mri = _MRIHeadWithFiducialsModel(subjects_dir=subjects_dir,
+                                               subject=subject)
+        self._nearest_calc = self._nearest_calc_default()
+
+    def _parameters_items_changed(self):
+        # Only update our nearest-neighbor if necessary
+        if self._parameters[6:9] != self._last_parameters[6:9]:
+            self._update_nearest_calc()
+        self._last_parameters[:] = self._parameters[:]
+
+    @property
+    def _changes(self):
+        new = np.array(self._parameters, float)
+        old = np.array(self._last_parameters, float)
+        move = np.linalg.norm(old[3:6] - new[3:6]) * 1e3
+        angle = np.rad2deg(_angle_between_quats(
+            rot_to_quat(rotation(*new[:3])[:3, :3]),
+            rot_to_quat(rotation(*old[:3])[:3, :3])))
+        percs = 100 * (new[6:] - old[6:]) / old[6:]
+        return move, angle, percs
+
+    @property
+    def _transformed_hsp_hpi(self):
+        return apply_trans(self._hsp_trans, self._hsp.hpi_points)
```

Comment:
We can make these properties, but the advantage of caching from `traitsui` goes away, which is sad. Some of these will end up being a bit slow to recompute, hence the recompute-only-when-necessary behavior in `main` is nice.

Reply:
Having this class separated out from the GUI stuff, I think the best way forward is to have a single `def _update_params(params)` method that looks at the `_last_params` and the new ones and updates the exact set of `_`-variables that need to be updated. This removes all the `property` business, and also makes it so that things can be updated in the right order (which was always a problem with the exsting `coreg` code actually!).

I suggest that the first step be the renaming ._hsp->._dig and DigSource.points - > DigSource.extra_points. Then the next step/commit is removal of all of these `@property`s using a single monolithic `_update_params` that does the right thing. (If you set for example `self._last_params = np.array([np.nan] * 9)` then call `self._update_params([0] * 6 + [1] * 3)` at the end of the class constructor, it should automatically populate all `_`-variables nicely.

Reply:
... and if this is not clear, I'm happy to have a video chat on Discord to discuss, or I can push some commits to move things in this direction!

Reply:
+100